node/src/components/linear_chain_sync.rs

```diff
@@ -147,53 +310,91 @@ where
-                        trace!(?init_hash, "Start synchronization");
+                    State::SyncingTrustedHash { trusted_hash, .. } => {
+                        trace!(?trusted_hash, "Start synchronization");
                         // Start synchronization.
-                        fetch_block(effect_builder, init_peer, init_hash)
+                        fetch_block_by_hash(effect_builder, init_peer, trusted_hash)
+                    }
+                    State::SyncingDescendants { .. } => {
+                        panic!("`Start` event received when in `SyncingDescendants` state.")
                     }
                 }
             }
-            Event::BlockExecutionDone(block_hash, block_height) => {
-                info!(
-                    ?block_hash,
-                    ?block_height,
-                    "Finished linear chain blocks execution."
-                );
-                Effects::new()
-            }
-            Event::GetBlockResult(block_hash, fetch_result) => match fetch_result {
+            Event::GetBlockHeightResult(block_height, fetch_result) => match fetch_result {
+                BlockByHeightResult::Absent => match self.random_peer(rng) {
+                    None => {
+                        // `block_height` not found on any of the peers.
+                        // We have synchronized all, currently existing, descendants of trusted
+                        // hash.
+                        effect_builder
+                            .immediately()
+                            .event(move |_| Event::StartDownloadingDeploys)
+                    }
+                    Some(peer) => fetch_block_at_height(effect_builder, peer, block_height),
+                },
+                BlockByHeightResult::FromStorage(block) => {
+                    self.state.block_downloaded(&block);
+                    // We should be checking the local storage for linear blocks before we
+                    // start syncing.
+                    trace!(%block_height, "Linear block found in the local storage.");
+                    // If we found the linear block in the storage it means we should have
+                    // all of its parents as well. If that's not
+                    // the case then we have a bug.
+                    effect_builder
+                        .immediately()
+                        .event(move |_| Event::StartDownloadingDeploys)
+                }
+                BlockByHeightResult::FromPeer(block, peer) => {
+                    self.state.block_downloaded(&block);
+                    if block.height() != block_height {
+                        warn!(
+                            "Block height mismatch. Expected {} got {} from {}.",
+                            block_height,
+                            block.height(),
+                            peer
+                        );
+                        // NOTE: Signal misbehaving validator to networking layer.
```

Comment:
Maybe also remove the peer from `self.peers`?

Same for misbehaviour in `Event::GetBlockHashResult()` below?

Reply:
I cannot do it for `GetBlockHashResult` b/c `peer` there has fixed type `NodeId/KeyFingerprint` :(

Reply:
Done.